Improve source management

guillep opened the pull request:
The main goal of this PR is to add optional meta-data into the source log files, with the main use case of storing debug metadata close to the source code, and not having it in memory when not needed.

The thing is, most of the code writing to files is duplicated and dispersed.
This happened because the underlying format (chunks) was the same, the usage was slightly different.
So the main work in this PR was removing duplications, clarifying APIs and responsibilities.

So with @Ducasse:

 - We refactored the fileout, `PositionalStream` and source/changes management to remove duplications => now chunk format is managed in a single place (the `CodeExporter` class)
 - We added `Fileouter`, a driver of `CodeExporter` that encapsulates the behavior of filing out
 - We added a new chunk after source code and comments to contain metadata. This tag is initially empty in this PR
 - We removed old support to parse `parseLangTagFor:`  that used to support multi-encoded strings which have not been used for years
 - We inlined the implementation of `RemoteString`, only used by the condenser, which otherwise made the refactoring very complicated

Also, during this pass, we made several quality of life improvements:
 - We improved the places in calypso and the debugger that use a fileout by adding a popup asking for the file to write to
 - We improved the move method refactoring, used by calypso's drag and drop, that was breaking in a border case when the order of the changes is important
 - We cleaned up usages of ChangesLog, or redirected duplicated code to ChangesLog when duplicated for nothing (e.g., Epicea)
 - Cleaning version metadata in `Smalltalk` et co
 - Improve method compilation by installing methods before announcing/categorizing/storing source. This way, if any of those break, the method is installed anyways. Otherwise, a bug there will make it impossible to recover.

guillep committed:
Cleaning PositionableStream extensions

guillep committed:
Cleaning version metadata in SmalltalkImage

guillep committed:
Remove dead code

guillep committed:
Use chunk writer

guillep committed:
Cleanup obsolete parseLangTagFor:

guillep committed:
Unify method access, chunk writing and method metadata

guillep committed:
Clean comment writing

guillep committed:
Extract chunk writer

guillep committed:
Unify code export

guillep committed:
Remove dead code

guillep committed:
Introduce fileouter

guillep committed:
Remove old fileout methods

guillep committed:
Improve code exporter API to write code for the first time

guillep committed:
Remove debugging code

guillep committed:
Cleaning ChangesLog and dependencies

guillep committed:
Fix drag and drop of methods: new method should be added before the old method is removed.
Otherwise this can leave the system in an invalid state

guillep committed:
Remove remote string

guillep committed:
Remove duplications in Behavior and subclasses

guillep committed:
(re)Add fileout growl

guillep committed:
Rework source and changes API

jecisc commented on src/CodeExport/CodeExporter.class.st:
```diff
@@ -4,11 +4,28 @@ An utility class for exporting code to changesets (*.cs) or Smalltalk source (*.
 	#tag : 'Base'
 }
 
+{ #category : 'as yet unclassified' }
```
```suggestion
{ #category : 'instance creation' }
```

jecisc commented on src/CodeExport/CodeExporter.class.st:
```diff
@@ -4,11 +4,28 @@ An utility class for exporting code to changesets (*.cs) or Smalltalk source (*.
+	^ self onStream: (String new: 100) writeStream
+]
+
+{ #category : 'as yet unclassified' }
```
```suggestion
{ #category : 'instance creation' }
```

guillep committed:
Better organizing dependencies

guillep committed:
Update src/CodeExport/CodeExporter.class.st

Co-authored-by: CyrilFerlicot <cyril@ferlicot.fr>

guillep committed:
Update src/CodeExport/CodeExporter.class.st

Co-authored-by: CyrilFerlicot <cyril@ferlicot.fr>

guillep committed:
Fixes:
 - writing comments should return the comment position
 - writing methods should take into account the metadata parameter

guillep committed:
Remove bugged bang

guillep committed:
Avoid metadata in comments as it breaks compatibility

jecisc commented:
> Apparently the issue is now
> 
> 
> 
> ```
> 
> SystemNotification: Loading System-Platforms-tonel.1.mcz
> 
> An error happened while reading MCZ. We will fallback to another format. Short error stack: ReadStream(Object)>>doesNotUnderstand: #nextWordsInto:
> 
> WideString class>>newFromStream:
> 
> MCDataStream>>readWordLike
> 
> MCDataStream>>next
> 
> MCClassDefinition(Object)>>readDataFrom:size:
> 
> MCDataStream>>readInstance
> 
> MCDataStream>>next
> 
> ```

This is not a problem.

The .mcz importer depends on graphic methods and night fail the import. 
So we fallback to importing the .st file. 

And later the graphic part will be loaded.

guillep committed:
Recover lost methods

guillep commented:
I was able to reproduce the issue from the duplicate class refactoring!

1. Browse `SpAbstractTreePresenter`
2. Right click, "duplicate". This opens the refactoring pop up dialog. 
3. Add a new name (like, suffix it with 2), accept. This opens a list of changes.
4. Accept the list of changes

=> boom, exception

<img width="1202" height="701" alt="imagen" src="https://github.com/user-attachments/assets/94cf01ef-354e-43c1-ba3c-d36a6159f6ca" />

guillep committed:
Recover lost method

guillep committed:
Classify before saving source. Source code requires protocol to be set.

guillep committed:
Removing a tag should remove the behaviors from the tag in topo order, not the behaviors of the package

guillep commented:
Now the PR is failing because there is some strange side effect with traits.
Reproducible using the following expression in the built image:

```smalltalk
HDTestReport
	shuffleSeed: 622728115;
 	runPackage: 'Traits-Tests'.

HDTestReport
	shuffleSeed: 588811568;
 	runPackage: 'MonticelloTonel-Tests'
```

guillep committed:
Fix trait removal: topological sort should only consider given traits and not others

guillep committed:
Recover method comments after move up

guillep committed:
Explicitly set protocol when available

guillep committed:
Fix epicea tests

guillep committed:
Fix unclassified methods

guillep committed:
Move lineIntervalAtCursorPosition up

guillep committed:
Source timestamp should be a string

guillep committed:
Convert stamp to string

guillep committed:
Fix testProtocolOfMethodRemoval.

Looks like protocols are set even for anonymous classes. This does not make much sense.

guillep committed:
Categorize method

guillep committed:
Stamps are strings. Fix tests

guillep committed:
Recover lost method

guillep committed:
Fix dependencies:

- implement a topological sort based on dependencies to avoid direct dependencies to traits
- move `ChunkReadStream` to `System-Sources``

guillep committed:
Install the method and classify AFTER setting the source pointer.
If the method is a trait method, trait installation is hooked down here.
And it depends on having set the source pointer before.
	
This is not the ideal order, because this means that a failure in writing the source code may prevent method installation in the system (and thus preventing further work/updates to code).

guillep committed:
Recover MethodChunk>>#methodSelector

guillep committed:
Categorize topo sort method

guillep committed:
Remove unused packages from baseline
